src/decode_url.rs

```diff
@@ -0,0 +1,62 @@
+use crate::encode_url::encode_url;
+use napi::bindgen_prelude::*;
+use std::borrow::Cow;
+
+#[inline]
+fn from_hex_digit(digit: u8) -> Option<u8> {
+  match digit {
+    b'0'..=b'9' => Some(digit - b'0'),
+    b'A'..=b'F' => Some(digit - b'A' + 10),
+    b'a'..=b'f' => Some(digit - b'a' + 10),
+    _ => None,
+  }
+}
+
+pub fn decode_url_impl(url: &str) -> Cow<[u8]> {
```

Comment:
Can we consume a `String` and return a `String` here? It will make this function less generic, but there will be less memory allocations and UTF-8 checks for regular URLs.